### ISSUE TITLE
TypeScript error when using async resolver predicates

### ISSUE THREAD

**Issue by kulla** [kettanaito/msw#1856]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 18 or higher

### Node.js version

v18.16.1

### Reproduction repository

https://github.com/kulla/2023-11-14-msw-examples

### Reproduction steps

* `npm install`
* `npm run test`

### Current behavior

I have added the example https://mswjs.io/docs/best-practices/custom-request-predicate#request-body-predicate into the example repo. Currenlty it fails with a typescript type error.

### Expected behavior

There should not be a type error when using async predicate functions.

**Comment by kulla** [comment]
## Reason for the bug

Let's take the following example:

```typescript
import { ResponseResolver } from 'msw'

function simplePredicate(resolver: ResponseResolver): ResponseResolver {
  return async (args) => resolver(args)
}
```

The return type of `resolver(args)` has the form `OtherTypes | Promise<PromiseTypes> | Generator<GeneratorTypes>` due to the current type definition of a `ResponseResolver`. Therefore `async (args) => resolver(args)` has the return type `Promise<OtherTypes | Promise<PromiseTypes> | Generator<GeneratorTypes>>` so it might be a promise of a generator.

This does not match the return type of `ResponseResolver`. It include promises, however in the current definition `PromiseTypes` cannot be a generator.

## Examples

At https://github.com/kulla/2023-11-14-msw-examples/blob/main/src/with-json-body.ts I have added some examples. By running `npm run test` one can see the type error in https://github.com/kulla/2023-11-14-msw-examples/blob/807e3036e3404c49dbb6ab7dda9596e2a2db0bb8/src/with-json-body.ts#L16.

I have added two examples how one can make typescript happy. It shows that the generator type is the reason for the type error.

**Comment by kettanaito** [comment]
Thanks for reporting this. This has been fixed in #1955 (#1961). Please read the [Using with TypeScript](https://mswjs.io/docs/best-practices/typescript) page in our docs to see the usage examples of MSW with TypeScript. Those include the higher-order response resolver too!

**Comment by kulla** [comment]
Nice, thank you! :heart:
